pages/docs/couple-your-code/couple-your-code-adapter-software-engineering.md

```diff
@@ -2,17 +2,119 @@
+Such encapsulation avoids duplicating additional checks and details and makes maintenance easier.
+
+This method is particularly useful for adapters used in frameworks that the user uses to create concrete simulations.
+Examples here include the [FEniCS adapter](./adapter-fenics.html) and the [deal.II adapter](https://github.com/precice/dealii-adapter).
+
+Even when directly modifying the code, it makes sense to hide as many changes as possible behind a class API.
+Examples for this include the [CalculiX adapter](./adapter-calculix-overview.html) and the [SU2 adapter for v6 and earlier](./adapter-su2-overview.html) (as described in [Alexander Rusch's thesis](https://www5.in.tum.de/pub/Rusch2016_BA.pdf)).
+
+Adapter classes can still be provided as optional modules of the main codebase. Examples include the [deal.II adapter](./adapter-dealii-overview.html), the [G+smo adapter](./adapter-gismo.html), the [DUNE adapter](./adapter-dune.html), and the [DuMux adapter](./adapter-dumux-get.html)
+
+### Adapter plugin
+
+Some solvers provide a plugin interface.
+Examples include OpenFOAM (function objects) and Fluent (UDF files).
+The solver exposes specific "hooks"/"callbacks", which trigger the execution of external code at specific stages of the simulation.
+Creating an adapter as such a plugin is non-intrusive and makes it easy to distribute the adapter independently.
+
+While this is a generally preferable architecture, it comes with some requirements.
+The solver needs to provide the following functionality via the plugin interface:
+
+- Executing additional steps during the configuration, setup, and teardown
+- Access to the (boundary) mesh coordinates (read)
+- Access to the (boundary) values (read/write)
+- Access to the time step size (read/write)
+- Controlling the exit condition of the time loop
+- Checkpointing:
+  - Storing and restoring the time
+  - Storing and restoring the internal solution
+
+The trickiest part is typically the checkpointing. However, in some cases, this can alternatively be implemented as (a) controlling when the solver moves to the next time step or discards the current solution and tries again, (b) storing restart files and restarting from them (inefficient, but might not be an issue in some cases).
+
+Examples of adapters implemented as plugins include the [OpenFOAM adapter](./adapter-openfoam-overview.html) (with the plugin approach explained in detail in [Gerasimos Chourdakis' thesis](https://www5.in.tum.de/pub/Chourdakis2017_Thesis.pdf)) and the [Fluent adapter](https://github.com/precice/fluent-adapter).
+
+### Adapter wrapper
+
+Some solvers provide an API, meaning that specific functions of the solver can be executed from external programs.
+A common example is a Python API, which allows us to write a Python wrapper for the solver.
+This wrapper then can call both the solver and preCICE, making the wrapper an adapter.
```

Comment:
We are in some gray area here: Also FEniCS has an API, as an example. Is an FEniCS application script then also an adapter? Not sure how to best solve this, but maybe we could write somewhere that these categories are ambiguous to some extent.